feat: Implements queuing full voice channels

hbina opened the pull request:
(Garbage) implementation of this feature.
Still trying to figure out how to do this sanely

See #7

Signed-off-by: Hanif Bin Ariffin <hanif.ariffin.4326@gmail.com>

hbina committed:
feat:Implements queuring full voice channels

(Garbage) implementation of this feature.
Still trying to figure out how to do this sanely

See #7

Signed-off-by: Hanif Bin Ariffin <hanif.ariffin.4326@gmail.com>

hbina committed:
Added some more functionality to this feature.

1. Now checks if the user is not empty.
2. Simplified implementation of workOnQueue.
3. Next, we need to check if user is _already_ in the channel.

Signed-off-by: Hanif Ariffin <hanif.ariffin.4326@gmail.com>

cglotr commented on src/managers/roommanagerimpl.ts:
```diff
@@ -81,4 +86,31 @@ export class RoomManagerImpl implements RoomManager {
   }
+
+  public addUserToQueue(user: Discord.User, channel: VoiceChannel) {
+    this.userQueue.push({ user, channel });
```
This will be potentially contain duplicates?

cglotr commented on src/types/channelqueue.ts:
```diff
@@ -0,0 +1,7 @@
+import { VoiceChannel } from './voicechannel';
+
+export type ChannelQueue = {
+  user: Discord.User;
```
I think better to only store the user id & fetch it on demand when we want to.

cglotr commented on src/main.ts:
```diff
@@ -340,6 +366,7 @@ client.on('voiceStateUpdate', (oldState, newState) => {
         newState.channel.members.size,
       );
     }
+    roomManager.workOnQueues();
```
I think we should move this to a cron process.

cglotr commented on src/managers/roommanagerimpl.ts:
```diff
@@ -81,4 +86,31 @@ export class RoomManagerImpl implements RoomManager {
+
+  public workOnQueues() {
+    this.loggerService.info(`queues:${JSON.stringify(this.userQueue)}`);
+    const nextQueue: ChannelQueue[] = [];
```
Should we have `n` number of queues, one for each tracked room?

hbina committed:
Reimplementation of issue#7

The implementation should be a lot cleaner and follows the
style of this repository.

Signed-off-by: Hanif Ariffin <hanif.ariffin.4326@gmail.com>

hbina commented:
@cglotr applied the requested changes. The only thing left is prolly checking edge cases and adding unit tests.

hbina commented on src/managers/queuemanagerimpl.test.ts:
```diff
@@ -0,0 +1,103 @@
+import { DiscordGuild } from '../discord/discordguild';
```
This is the _wrong_ test.

hbina committed:
Further implementation of the requested feature.

Added unit tests.

Signed-off-by: Hanif Ariffin <hanif.ariffin.4326@gmail.com>
